components/remerge/src/schema/json.rs

```diff
@@ -0,0 +1,1178 @@
+
+//! This module is concerned primarially with schema parsing (from RawSchema,
+//! e.g. the schema represented as JSON), and validation. It's a little bit
+//! hairy, and for the definitive documentation, you should refer to the
+
+// Clippy seems to be upset about serde's output:
+// https://github.com/rust-lang/rust-clippy/issues/4326
+#![allow(clippy::type_repetition_in_bounds)]
+
+use super::desc::*;
+use super::error::*;
+use super::merge_kinds::*;
+use crate::util::is_default;
+use index_vec::IndexVec;
+use matches::matches;
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+use std::collections::{HashMap, HashSet};
+use url::Url;
+
+pub const FORMAT_VERSION: usize = 1;
+
+pub fn parse_from_string(json: &str, is_remote: bool) -> Result<RecordSchema, SchemaError> {
+    let schema = match serde_json::from_str::<RawSchema>(json) {
+        Ok(schema) => schema,
+        Err(e) => {
+            // If it's local then this is just a format error.
+            if !is_remote {
+                // For some reason throw! and ensure! both complain about moving
+                // `e` here, but this works...
+                return Err(SchemaError::FormatError(e));
+            }
+            // If it's remote, then it failed, but it could have failed because
+            // it's from a future version. Check that.
+            let version = match serde_json::from_str::<JustFormatVersion>(json) {
+                Ok(s) => s.format_version,
+                Err(_) => {
+                    // Ditto with moving `e` (which we want to use because it can give
+                    // better error messages).
+                    return Err(SchemaError::FormatError(e));
+                }
+            };
+            return Err(if version != FORMAT_VERSION {
+                SchemaError::WrongFormatVersion(version)
+            } else {
+                SchemaError::FormatError(e)
+            });
+        }
+    };
+    let parser = SchemaParser::new(&schema, false);
+    Ok(parser.parse()?)
+}
+
+/// Helper trait to make marking results / errors with which field were were
+/// parsing more convenient.
+trait FieldErrorHelper {
+    type Out;
+    fn named(self, name: &str) -> Self::Out;
+}
+
+impl FieldErrorHelper for FieldError {
+    type Out = SchemaError;
+    fn named(self, name: &str) -> SchemaError {
+        SchemaError::FieldError(name.into(), self)
+    }
+}
+
+impl<T> FieldErrorHelper for Result<T, FieldError> {
+    type Out = Result<T, SchemaError>;
+    fn named(self, name: &str) -> Result<T, SchemaError> {
+        match self {
+            Ok(v) => Ok(v),
+            Err(e) => Err(e.named(name)),
+        }
+    }
+}
+
+// Used just to parse out the required version in case JsonSchema changes incompatibly.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct JustFormatVersion {
+    format_version: usize,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RawSchema {
```

Comment:
The duplication between these and everything in `desc.rs` is unfortunate. It may be worth expanding the comments to mention that, if you change the schema format in any way, you need to make the change in both places, and also add validation code here.